### PR TITLE
Conditional Death

### PR DESCRIPTION
**Pull request by FS-21**
New ways for self-killing objects under certaing cases.
· No Ammo: The object will die if the remaining ammo reaches 0.
· Countdown: The object will die if the countdown (in frames) reaches 0. Optionally the affected object can decrease the health until the countdown ends.

In rulesmd.ini:

[SOMETECHNO]                       ; TechnoType
Death.NoAmmo=no                    ; boolean
Death.Countdown=0                  ; integer
Death.Countdown.DecreaseHealth=no  ; boolean

### CODE REVIEW

File: src/Ext/Techno/Body.cpp [hunk 1]
```diff
@@ -247,6 +247,55 @@ CoordStruct TechnoExt::GetBurstFLH(TechnoClass* pThis, int weaponIndex, bool& FL
...
+	auto pTypeData = TechnoTypeExt::ExtMap.Find(pTypeThis);
+	auto pData = TechnoExt::ExtMap.Find(pThis);
+
+	// Death if no ammo
+	if (pTypeThis && pTypeData && pTypeData->DeathIfNoAmmo)
+	{
+		if (pTypeThis->Ammo > 0 && pThis->Ammo <= 0)
+		{
+			pThis->ReceiveDamage(&pThis->Health, 0, RulesClass::Instance()->C4Warhead, nullptr, true, false, pThis->Owner);
+		}
```
Comment (YRIDZE) [inline src/Ext/Techno/Body.cpp:263]:
```suggestion
		if (pTypeThis->Ammo > 0 && pThis->Ammo <= 0)
			pThis->ReceiveDamage(&pThis->Health, 0, RulesClass::Instance()->C4Warhead, nullptr, true, false, pThis->Owner);
```

File: src/Ext/Techno/Body.cpp [hunk 2]
```diff
@@ -247,6 +247,55 @@ CoordStruct TechnoExt::GetBurstFLH(TechnoClass* pThis, int weaponIndex, bool& FL
...
+		{
+			if (pData->DeathIfCountdown > 0)
+			{
+				pData->DeathIfCountdown--; // Update countdown
+
+				if (pTypeData->DeathIfCountdown_DecreaseHealth)
+				{
+					// Use the Health bar as a kind of indicator (in the case of immortal units, if not it simply weakens the unit)
+					double multiplier = ((pData->DeathIfCountdown * 100.0) / pTypeData->DeathIfCountdown) / 100.0;
+
```
Comment (YRIDZE) [inline src/Ext/Techno/Body.cpp:279]:
```suggestion
```

File: src/Ext/Techno/Body.cpp [hunk 3]
```diff
@@ -247,6 +247,55 @@ CoordStruct TechnoExt::GetBurstFLH(TechnoClass* pThis, int weaponIndex, bool& FL
...
+					if (pThis->Health > 0 && pThis->Health > newHP && newHP > 0)
+						pThis->Health = newHP;
+				}
+			}
+			else
+			{
+				// Countdown ended. Kill the unit
+				pData->DeathIfCountdown = -1;
+
```
Comment (YRIDZE) [inline src/Ext/Techno/Body.cpp:290]:
```suggestion
```

File: src/Ext/Techno/Body.cpp [hunk 4]
```diff
@@ -247,6 +247,55 @@ CoordStruct TechnoExt::GetBurstFLH(TechnoClass* pThis, int weaponIndex, bool& FL
...
+				pThis->ReceiveDamage(&pThis->Health, 0, RulesClass::Instance()->C4Warhead, nullptr, true, false, pThis->Owner);
+			}
+		}
+		else
+			pData->DeathIfCountdown = pTypeData->DeathIfCountdown; // Start countdown
```
Comment (YRIDZE) [inline src/Ext/Techno/Body.cpp:295]:
```suggestion
		else
		{
			pData->DeathIfCountdown = pTypeData->DeathIfCountdown; // Start countdown
		}
```

File: docs/New-or-Enhanced-Logics.md [hunk 5]
```diff
@@ -300,6 +300,20 @@ WeaponXFLH.BurstN=             ; int - forward, lateral, height
 EliteWeaponXFLH.BurstN=        ; int - forward, lateral, height
 ```
 
+### Kill Unit Automatically
+
+- New ways for self-killing objects under certaing cases.
```
Comment (YRIDZE) [inline docs/New-or-Enhanced-Logics.md:305]: TechnoTypes can now self-destruct in new ways in certain cases.

**Review by YRIDZE — DISMISSED** [review]
Code looks good except for some braces and newlines. Please, fix it and add info to Whats-New.md

**Comment by CommanderNour** [comment]
Tested, Everything works just fine

**Comment by FS-21** [comment]
Thanks for the testing :-)
This PR can be reviewed if necessary & if acepted then merge.

File: src/Ext/Script/Body.cpp [hunk 6]
```diff
@@ -143,6 +143,15 @@ void ScriptExt::ProcessAction(TeamClass* pTeam)
...
 		break;
+	default:
+		// Do nothing because or it is a wrong Action number or it is an Ares/YR action...
+		if (action > 70 && !(action >= PhobosScripts::LocalVariableAdd && action <= PhobosScripts::GlobalVariableAndByGlobal))
+		{
+			// Unknown new action. This action finished
+			pTeam->StepCompleted = true;
+			Debug::Log("[%s] [%s] (line %d): Unknown Script Action: %d\n", pTeam->Type->ID, pTeam->CurrentScript->Type->ID, pTeam->CurrentScript->idxCurrentLine, pTeam->CurrentScript->Type->ScriptActions[pTeam->CurrentScript->idxCurrentLine].Action);
+		}
+		break;
```
Comment (Metadorius) [inline src/Ext/Script/Body.cpp:154]: This doesn't belong here.

File: src/Ext/Techno/Body.cpp [hunk 7]
```diff
@@ -385,6 +385,53 @@ bool TechnoExt::CanFireNoAmmoWeapon(TechnoClass* pThis, int weaponIndex)
 	return false;
 }
 
+// Feature: Kill Object Automatically
+void TechnoExt::ApplyDeath_If(TechnoClass* pThis)
```
Comment (Metadorius) [inline src/Ext/Techno/Body.cpp:389]: This function name is really broken, `CheckDeathConditions` would be better.

File: docs/New-or-Enhanced-Logics.md [hunk 8]
```diff
@@ -374,6 +374,20 @@ In `rulesmd.ini`:
...
+- No Ammo: The object will die if the remaining ammo reaches 0.
+- Countdown: The object will die if the countdown reaches 0. Optionally the affected object can decrease the Health until the countdown ends.
+
+In `rulesmd.ini`:
+```ini
+[SOMETECHNO]                       ; TechnoType
+Death.NoAmmo=no                    ; boolean
+Death.Countdown=0                  ; integer
+Death.Countdown.DecreaseHealth=no  ; boolean
```
Comment (Metadorius) [inline docs/New-or-Enhanced-Logics.md:388]: Mismatch between tag names and internal variable names.

File: docs/New-or-Enhanced-Logics.md [hunk 9]
```diff
@@ -374,6 +374,20 @@ In `rulesmd.ini`:
 NoSecondaryWeaponFallback=false   ; boolean
 ```
 
+### Kill Unit Automatically
+
+- New ways for self-killing objects under certaing cases.
+- No Ammo: The object will die if the remaining ammo reaches 0.
+- Countdown: The object will die if the countdown reaches 0. Optionally the affected object can decrease the Health until the countdown ends.
```
Comment (Metadorius) [inline docs/New-or-Enhanced-Logics.md:381]: The `decrease health` part looks pretty pointless to me, one can make the same stuff using a weapon that techno is firing on itself.

File: docs/New-or-Enhanced-Logics.md (hunk 5 above)
Comment (Metadorius) [reply docs/New-or-Enhanced-Logics.md:379]: Brambuca meant that this line should be rephrased.

**Review by Metadorius — CHANGES_REQUESTED** [review]

**Review by Metadorius — APPROVED** [review]
Fix conflicts and merge.

**Pull request merged**